grep-0001-coding-guidelines.md

```diff
@@ -0,0 +1,122 @@
+
+New code should follow these rules, but there are always exceptions, and it is
+up to the individual contributor to decide what works best. That said, certain
+things such as indentation or whitespace rules can be taken care of by one's
+editor, and there is no reason to not adopt these settings. This makes it
+easier for multiple people to contribute to the same codebase, and will keep
+commits free of non-functional changes.
+
+To summarize, consistency is extremely useful, and coding guidelines help ensure
+consistency. We assume that all contributors are mature enough to at least try
+and follow them.
+
+## Description
+
+### General Coding Guidelines
+
+* Never submit code with trailing whitespace.
+* Code is read more often than it's written. Code readability is thus something
+  worth optimizing for.
+* Try and keep line lengths to 79 characters, unless readability suffers. We
+  often have to do fun things like SSH into machines and edit code in a
+  terminal, and do side-by-side views of code on small-ish screens, so this is
+  actually pretty helpful.
+
+### C++-specific Guidelines
+
+* If in doubt, consult the [C++ Core Guidelines][https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md].
+  If the guidelines have an answer, and it works for you, just pick that.
+* Use Doxygen doc-blocks copiously.
+* All things equal, prefer standard C++ constructs over Boost constructs (see
+  also Boost guidelines).
+* Given the option, prefer C++ lambdas over std::bind, and just don't use
+  boost::bind if you can.
+* Include include files in the following order: Local headers, other GNU Radio
+  headers, 3rd-party library headers, Boost headers, standard headers.
+  The rationale is to include from most to least specific. This is the best way
+  to catch missing includes (if you were to include the standard header first,
+  it would be available to all include files that come later. If they need that
+  standard header too, they should be including it themselves).
+  Example:
+
+```cpp
+#include "module_specific.hpp"
+#include <gnuradio/block.h>
+#include <fftw3.h>
+#include <boost/shared_ptr.hpp>
+#include <mutex>
+```
+
+* Feel free to use modern C/C++ features even if they were not used before.
+  Make sure they work with the compilers and dependencies which are set for the
+  version of UHD the commit will be made upon. The GNU Radio CI system will be
+  able to confirm this. (Note: C++11 features are available starting with
+  version 3.8).
+
+
+## Boost-specific Guidelines
+
+* Avoid Boost where possible.
+* Don't use Boost's sleep functions. Prefer `std::chrono` functions.
+
+## Python-specific Guidelines
+
+* Follow the suggestions in PEP8 (https://www.python.org/dev/peps/pep-0008/)
+  and PEP257 (https://www.python.org/dev/peps/pep-0257/). The former is about
+  code layout in general, the latter about docstrings.
+* Keep Python code compatible with Py2k and Py3k, where possible.
```

Comment:
I assume it's better to be explicit and set a standard for the supported Python version, which in 2.X should be 2.7 IMO.

Reply:
That's a good point; we actually need to lock down Python versions actually. For the upcoming 3.8 release, we might just drop Py2k support altogether and then this point needs some rewriting anyway.

Reply:
Yes, I'd argue that this document would apply to 3.8 and further on, anyway. Py2k statements might optimally be found under "if you're changing lecacy (pre-3.8) code…", or so?